tests/withBlockTracker.ts

```diff
@@ -0,0 +1,252 @@
+  };
+}
+
+interface WithSubscribeBlockTrackerTypes {
+  withBlockTrackerCallback: (args: {
+    provider: FakeProvider;
+    blockTracker: SubscribeBlockTracker;
+  }) => void | Promise<void>;
+  withBlockTrackerOptions: {
+    provider?: FakeProviderOptions;
+    blockTracker?: SubscribeBlockTrackerOptions;
+  };
+}
+
+type FakeProviderStub =
+  | {
+      methodName: string;
+      response: { result: any } | { error: string };
+    }
+  | {
+      methodName: string;
+      implementation: () => void;
+    }
+  | {
+      methodName: string;
+      error: string;
+    };
+
+interface FakeProviderOptions {
+  stubs?: FakeProviderStub[];
+}
+
+/**
+ * FakeProvider is an implementation of the provider that a subclass of
+ * BaseBlockTracker takes, supporting the same expected methods with the same
+ * expected interface, except that fake responses for the various RPC methods
+ * that the provider supports can be supplied. This ends up easier to define
+ * than using `jest.spyOn(...).mockImplementationOnce(...)` due to the types
+ * that `sendAsync` takes.
+ */
+class FakeProvider extends SafeEventEmitter implements Provider {
+  private stubs: FakeProviderStub[];
+
+  private originalStubs: FakeProviderStub[];
+
+  constructor({ stubs = [] }: FakeProviderOptions = {}) {
+    super();
+    this.stubs = this.buildStubsFrom(stubs);
+    this.originalStubs = this.stubs.slice();
+  }
+
+  sendAsync<T, U>(
+    request: JsonRpcRequest<T>,
+    callback: (err: Error, response: JsonRpcResponse<U>) => void,
+  ) {
+    const index = this.stubs.findIndex(
+      (stub) => stub.methodName === request.method,
+    );
+
+    if (index !== -1) {
+      const stub = this.stubs[index];
+      this.stubs.splice(index, 1);
+      if ('implementation' in stub) {
+        stub.implementation();
+      } else if ('response' in stub) {
+        if ('result' in stub.response) {
+          callback(null as unknown as Error, {
+            jsonrpc: '2.0',
+            id: 1,
+            result: stub.response.result,
+          });
+        } else if ('error' in stub.response) {
+          callback(null as unknown as Error, {
+            jsonrpc: '2.0',
+            id: 1,
+            error: {
+              code: -999,
+              message: stub.response.error,
+            },
+          });
+        }
+      } else if ('error' in stub) {
+        callback(new Error(stub.error), null as unknown as JsonRpcResponse<U>);
+      }
+      return;
+    }
+
+    callback(
+      new Error(
+        `Could not find any stubs matching "${request.method}". Perhaps they've already been called?\n\n` +
+          'The original set of stubs were:\n\n' +
+          `${util.inspect(this.originalStubs, { depth: null })}\n\n` +
+          'Current set of stubs:\n\n' +
+          `${util.inspect(this.stubs, { depth: null })}\n\n`,
+      ),
+      null as unknown as JsonRpcResponse<U>,
+    );
+  }
+
+  private buildStubsFrom(givenStubs: FakeProviderStub[]): FakeProviderStub[] {
```

Comment:
Nit: We can use an ECMA private field here instead (`#buildStubsFrom`)

Reply:
Good point! I also used private fields for all of the other private things. Changed in [c3e20c9](https://github.com/MetaMask/eth-block-tracker/pull/103/commits/c3e20c989f39fad72cc1bc2f002839d6db4f61cd).